src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonIEnumerableWithAddMethodConverter.cs

```diff
@@ -0,0 +1,82 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections;
+using System.Diagnostics;
+
+namespace System.Text.Json.Serialization.Converters
+{
+    internal sealed class JsonIEnumerableWithAddMethodConverter<TCollection> :
+        JsonIEnumerableDefaultConverter<TCollection, object?>
+        where TCollection : IEnumerable
+    {
+        protected override void Add(object? value, ref ReadStack state)
+        {
+            Debug.Assert(state.Current.ReturnValue is TCollection);
+            Debug.Assert(state.Current.AddMethodDelegate != null);
+            ((Action<TCollection, object?>)state.Current.AddMethodDelegate)((TCollection)state.Current.ReturnValue!, value);
+        }
+
+        protected override void CreateCollection(ref ReadStack state, JsonSerializerOptions options)
+        {
+            JsonClassInfo.ConstructorDelegate? constructorDelegate = state.Current.JsonClassInfo.CreateObject;
+
+            if (constructorDelegate == null)
+            {
+                ThrowHelper.ThrowNotSupportedException_DeserializeNoParameterlessConstructor(TypeToConvert);
+            }
+
+            state.Current.ReturnValue = constructorDelegate();
+            state.Current.AddMethodDelegate = GetAddMethodDelegate(options);
+        }
+
+        protected override bool OnWriteResume(Utf8JsonWriter writer, TCollection value, JsonSerializerOptions options, ref WriteStack state)
+        {
+            IEnumerator enumerator;
+            if (state.Current.CollectionEnumerator == null)
+            {
+                enumerator = value.GetEnumerator();
+                if (!enumerator.MoveNext())
+                {
+                    return true;
+                }
+            }
+            else
+            {
+                enumerator = state.Current.CollectionEnumerator;
+            }
+
+            JsonConverter<object?> converter = GetElementConverter(ref state);
+            do
+            {
+                if (ShouldFlush(writer, ref state))
+                {
+                    state.Current.CollectionEnumerator = enumerator;
+                    return false;
+                }
+
+                if (!converter.TryWrite(writer, enumerator.Current!, options, ref state))
```

Comment:
Is this null-forgiving on `enumerator.Current` necessary?

Reply:
Not necessary. Not sure why it was added. Did any rules regarding generic change in the last few weeks?